allennlp/models/semantic_parsing/text2sql_parser.py

```diff
@@ -0,0 +1,399 @@
+from allennlp.state_machines.states import GrammarBasedState
+from allennlp.state_machines.transition_functions import BasicTransitionFunction
+from allennlp.state_machines import BeamSearch
+from allennlp.state_machines.trainers import MaximumMarginalLikelihood
+from allennlp.state_machines.states import GrammarStatelet, RnnStatelet
+from allennlp.training.metrics import Average
+
+logger = logging.getLogger(__name__)  # pylint: disable=invalid-name
+
+
+@Model.register("text2sql_parser")
+class Text2SqlParser(Model):
+    """
+    Parameters
+    ----------
+    vocab : ``Vocabulary``
+    utterance_embedder : ``TextFieldEmbedder``
+        Embedder for utterances.
+    action_embedding_dim : ``int``
+        Dimension to use for action embeddings.
+    encoder : ``Seq2SeqEncoder``
+        The encoder to use for the input utterance.
+    decoder_beam_search : ``BeamSearch``
+        Beam search used to retrieve best sequences after training.
+    max_decoding_steps : ``int``
+        When we're decoding with a beam search, what's the maximum number of steps we should take?
+        This only applies at evaluation time, not during training.
+    input_attention: ``Attention``
+        We compute an attention over the input utterance at each step of the decoder, using the
+        decoder hidden state as the query.  Passed to the transition function.
+    database_file: ``str``, required.
+        The path of the SQLite database when evaluating SQL queries. SQLite is disk based, so we need
+        the file location to connect to it.
+    add_action_bias : ``bool``, optional (default=True)
+        If ``True``, we will learn a bias weight for each action that gets used when predicting
+        that action, in addition to its embedding.
+    dropout : ``float``, optional (default=0)
+        If greater than 0, we will apply dropout with this probability after all encoders (pytorch
+        LSTMs do not apply dropout to their last layer).
+    """
+    def __init__(self,
+                 vocab: Vocabulary,
+                 utterance_embedder: TextFieldEmbedder,
+                 action_embedding_dim: int,
+                 encoder: Seq2SeqEncoder,
+                 decoder_beam_search: BeamSearch,
+                 max_decoding_steps: int,
+                 input_attention: Attention,
+                 database_file: str,
+                 add_action_bias: bool = True,
+                 training_beam_size: int = None,
+                 dropout: float = 0.0,
+                 initializer: InitializerApplicator = InitializerApplicator(),
+                 regularizer: Optional[RegularizerApplicator] = None) -> None:
+        super().__init__(vocab, regularizer)
+
+        self._utterance_embedder = utterance_embedder
+        self._encoder = encoder
+        self._max_decoding_steps = max_decoding_steps
+        self._add_action_bias = add_action_bias
+        self._dropout = torch.nn.Dropout(p=dropout)
```

Comment:
Does this work when dropout is 0?  If it does, why have we been doing the `lambda x: x` stuff this whole time?